rllib/contrib/bandits/agents/linear.py

```diff
@@ -0,0 +1,159 @@
+UCB_PATH = "ray.rllib.contrib.bandits.exploration.UCB"
+
+
+class BanditPolicyOverrides:
+    @override(TorchPolicy)
+    def learn_on_batch(self, postprocessed_batch):
+        train_batch = self._lazy_tensor_dict(postprocessed_batch)
+        unflattened_obs = restore_original_dimensions(
+            train_batch[SampleBatch.CUR_OBS], self.observation_space,
+            self.framework)
+
+        info = {}
+
+        start = time.time()
+        self.model.partial_fit(unflattened_obs,
+                               train_batch[SampleBatch.REWARDS],
+                               train_batch[SampleBatch.ACTIONS])
+
+        infos = postprocessed_batch["infos"]
+        if "regret" in infos[0]:
+            regret = sum(
+                row["infos"]["regret"] for row in postprocessed_batch.rows())
+            self.regrets.append(regret)
+            info["cumulative_regret"] = sum(self.regrets)
+        else:
+            if log_once("no_regrets"):
+                logger.warning("The env did not report `regret` values in "
+                               "its `info` return, ignoring.")
+        info["update_latency"] = time.time() - start
+        return {LEARNER_STATS_KEY: info}
+
+
+def make_model_and_action_dist(policy, obs_space, action_space, config):
+    dist_class, logit_dim = ModelCatalog.get_action_dist(
+        action_space, config["model"], framework="torch")
+    model_cls = DiscreteLinearModel
+
+    if hasattr(obs_space, "original_space"):
+        original_space = obs_space.original_space
+    else:
+        original_space = obs_space
+
+    exploration_config = config.get("exploration_config")
+    # Model is dependent on exploration strategy because of its implicitness
+
+    # TODO: Have a separate model catalogue for bandits
+    if exploration_config:
+        if exploration_config["type"] == TS_PATH:
+            if isinstance(original_space, spaces.Dict):
+                assert "item" in original_space.spaces, \
+                    "Cannot find 'item' key in observation space"
+                model_cls = ParametricLinearModelThompsonSampling
+            else:
+                model_cls = DiscreteLinearModelThompsonSampling
+        elif exploration_config["type"] == UCB_PATH:
+            if isinstance(original_space, spaces.Dict):
+                assert "item" in original_space.spaces, \
+                    "Cannot find 'item' key in observation space"
+                model_cls = ParametricLinearModelUCB
+            else:
+                model_cls = DiscreteLinearModelUCB
+
+    model = model_cls(
+        obs_space,
+        action_space,
+        logit_dim,
+        config["model"],
+        name="LinearModel")
+    return model, dist_class
+
+
+def init_cum_regret(policy, *args):
+    policy.regrets = []
+
+
+# Build a common policy for TS and UCB
+LinUCBTSPolicy = build_torch_policy(
+    name="LinUCBTSPolicy",
+    get_default_config=lambda: DEFAULT_CONFIG,
+    loss_fn=None,
```

Comment:
Should we add a `validate_config=validate_config_fn` here?
Or would it make sense to run these with explorations different from UCB/TS?

Reply:
It is possible to use this policy with any exploration config. It does make sense as one might want to compare e-greedy or softmax sampling with UCB or TS.